certora/specs/SocialRecoveryModule.spec

```
@@ -1,26 +1,210 @@
+
+    assert safeContract.isModuleEnabled(currentContract);
+    assert e.msg.sender == safeContract;
+    assert !currentContract.isGuardian(safeContract, guardian);
+    assert guardianStorageContract.entries[safeContract].guardians[prevGuardian] == nextGuardian;
+    assert guardian != otherAccount => otherAccountIsGuardian == currentContract.isGuardian(safeContract, otherAccount);
+    assert currentGuardiansCount - 1 == to_mathint(guardianStorageContract.entries[safeContract].count);
+}
+
+// This integrity rule verifies that the guardian can always be revoked considering ideal conditions.
+rule guardianCanAlwaysBeRevoked(env e, address guardian, address prevGuardian, uint256 threshold) {
     requireSocialRecoveryModuleEnabled();
 
-    safeContract.disableModule@withrevert(e, prevModule, currentContract);
+    // No value should be sent with the transaction.
+    require e.msg.value == 0;
+    // Guardian should not be zero address.
+    require guardian != 0;
+    //New threshold should be greater than 0.
+    require threshold > 0;
+    // The guardian count should be greater than the new threshold.
+    require guardianStorageContract.countGuardians(safeContract) > threshold;
+    // The address should be a guardian.
+    require currentContract.isGuardian(safeContract, guardian);
+    require guardianStorageContract.isGuardianInGuardianList(safeContract, guardian);
+    requireGuardiansLinkedListIntegrity();
+
+    address nextGuardian = guardianStorageContract.entries[safeContract].guardians[guardian];
+    require guardianStorageContract.entries[safeContract].guardians[prevGuardian] == guardian;
+
+    uint256 currentGuardiansCount = guardianStorageContract.entries[safeContract].count;    
+
+    // Safe Contract should be the sender of the transaction.
+    require e.msg.sender == safeContract;
+    currentContract.revokeGuardianWithThreshold@withrevert(e, safeContract, prevGuardian, guardian, threshold);
+    bool isReverted = lastReverted;
+
+    assert !isReverted &&
+        guardianStorageContract.entries[safeContract].guardians[prevGuardian] == nextGuardian &&
+        !currentContract.isGuardian(safeContract, guardian) &&
+        currentGuardiansCount - 1 == to_mathint(guardianStorageContract.entries[safeContract].count);
+}
+
+// This integrity rule verifies the possibilites in which the revocation of a new guardian can revert.
+rule revokeGuardianRevertPossibilities(env e, address otherAccount, address prevGuardian, uint256 threshold) {
+    bool isNotGuardian = !currentContract.isGuardian(safeContract, otherAccount);
+
+    currentContract.revokeGuardianWithThreshold@withrevert(e, safeContract, prevGuardian, otherAccount, threshold);
     bool isReverted = lastReverted;
 
-    assert !isReverted => !safeContract.isModuleEnabled(currentContract);
+    assert isNotGuardian => isReverted;
+    assert !isNotGuardian && isReverted =>
+        e.msg.sender != safeContract ||
```

Is there a reason you can't write:

```suggestion
    assert isReverted =>
        isNotGuardian ||
        e.msg.sender != safeContract ||
```

---

(I remember that association direction has slightly different semantics, but I can't think of a counter example at the moment).

---

Actually, I wonder if the order of association here is correct.

I think, if we want to prove that it _always_ reverts, then we need to change the direction here.

```
assert isNotGuardian => isReverted;
assert e.msg.value != 0 => isReverted;
// etc.
```

---

Updated in both revert possibility rules 👍🏾